### DISCUSSION TITLE
Running homestar-runtime in a web browser

### DISCUSSION THREAD

**Discussion by andrewzhurov** [ipvm-wg/homestar#244] [General]
Will it be possible to run homestar-runtime in a web browser?
So it becomes possible to build web apps that leverage distributed compute / that are built on top of this IPOS.

**Comment by zeeshanlakhani** [comment]
@andrewzhurov This is on the agenda, without a doubt. 

- a) We'll have a JS runtime that can execute wasm-component-oriented workflows directly in the browser (if online to fetch wasm mods or if it has them cached, etc), or relay tasks/workflows to native nodes over the network if combined with other task types it can't directly handle, for example. 

- b) We're also exploring a UI to be deployed/released with Homestar that can relay workflows (via JSON-RPC, for example) to connected Homestar nodes on the backend, plus other need bits. 

I'm working on an example app that showcases some of the potential of *b* that people can spin up. It's not the exact interface, but we'll be developing both the CLI and web interfaces more thoroughly going forward. 

Hope that answers your Q :).

**Reply by andrewzhurov** [reply]
Thank you for a detailed and prompt reply! And for all the effort put into R&D of IPVM. :muscle: 
The idea of running IPVM in browser is exciting, neat that it's already in development.
On a side note, I'm intrigued [whether Rhizome could be integrated with IPVM](https://github.com/RhizomeDB/rs-rhizome/discussions/106). Would love to learn your thoughts on it, if it sparks interest in you as well.:)

**Reply by zeeshanlakhani** [reply]
@andrewzhurov Oh, I think there's a long-view vision of Rhizome + Ipvm + Wnfs all working together, as they all operate within the greater Fissionverse. For integration between Ipvm and Rhizome, I'm not sure when that will happen, but it's something we've discussed generally without a doubt.

**Reply by expede** [reply]
Yeah, we definitely want all the parts to work together. They're built in layers that get composed by the SDK, but designed to be interoperable and to take advanatge of each other's functionality (e.g. storing Rhizome in private WNFS, using IPVM to build indexes for Rhizome data, or build thumbnail images on WNFS changes, etc)

Here's an stack diagram of the big picture (with outdated names, e.g. WebNative Database = Rhizome)

<img width="955" alt="Screenshot 2023-08-17 at 10 57 01" src="https://github.com/ipvm-wg/homestar/assets/1052016/85ebc38e-3d52-4b69-988b-83dbde5ce684">
